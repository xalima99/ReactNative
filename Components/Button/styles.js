import { StyleSheet } from "react-native";
import { ColoredBackGround, ClearBackGround } from "../../constants/Colors";

const styles = StyleSheet.create({
  //container styles
  containerDefault: {
    alignItems: "center",
    justifyContent: "center",
    // alignSelf: 'center',
    paddingHorizontal: 16,
    borderRadius: 6,
    minHeight: 48,
    height: 48,
    marginVertical: 4,
    shadowColor: "#333",
    shadowOffset: {
      width: 0,
      height: .3,
    },
    shadowOpacity: 0.25,
    shadowRadius: 0.44,

    elevation: 0,
  },
  block: {
    minWidth: "100%",
  },
  containerPrimary: {
    backgroundColor: ClearBackGround.Fills.Default,
    borderColor: ClearBackGround.Fills.Default,
  },
  containerPrimaryOutlined: {
    backgroundColor: "transparent",
    borderWidth: 1,
    color: ClearBackGround.Fills.Default,
  },
  containerSecondary: {
    backgroundColor: ColoredBackGround.Fills.Default,
    borderColor: ColoredBackGround.Fills.Default,
  },
  containerSecondaryOutlined: {
    backgroundColor: "transparent",
    borderWidth: 1,
    // borderColor: ClearBackGround.Fills.Default,
  },

  containerDisabledPrimary: {
    backgroundColor: ClearBackGround.Fills.Disable,
    borderColor: ClearBackGround.Fills.Disable,
  },
  containerDisabledSecondary: {
    backgroundColor: ColoredBackGround.Fills.Disable,
    borderColor: ColoredBackGround.Fills.Disable,
  },
  containerDisabledOutlined: {
    borderColor: ColoredBackGround.Fills.Disable,
    backgroundColor: "transparent",
  },
  link: {
    borderWidth: 0,
  },
  linkPrimary: {
    borderWidth: 0,
    color: ClearBackGround.Fills.Default,
  },
  //text Styles
  textDefault: {
    fontSize: 16,
    fontWeight: "500",
    color: "#fff",
    // alignSelf: 'center'
  },
  textPrimary: {
    color: ClearBackGround.TextAndIcon.Default,
  },
  textPrimaryOutlined: {
    color: ClearBackGround.Fills.Default,
  },
  textSecondary: {
    color: ColoredBackGround.TextAndIcon.Default,
  },
  textSecondaryOutlined: {
    color: ColoredBackGround.Fills.Default,
  },
  textDisabledPrimary: {
    color: ClearBackGround.TextAndIcon.Disable,
  },
  textDisabledSecondary: {
    color: ClearBackGround.TextAndIcon.Disable,
  },
  textDisableOutlined: {
    color: ColoredBackGround.TextAndIcon.Disable,
  },

  //Container
  main: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  icon: {
    maxHeight: 20,
    paddingLeft: 16,
    maxWidth: 16,
    marginRight: 8,
  },
});

export default styles;
