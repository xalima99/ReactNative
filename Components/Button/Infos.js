import React from "react";
import { View, Image, Text } from "react-native";
import styles from "./styles";
import { ClearBackGround } from "../../constants/Colors";

const Infos = ({
  children,
  icon,
  iconColor,
  textStyles,
  secondary,
  outlined,
  isPressed,
}) => {
  return (
    <View style={styles.main}>
      {icon && (
        <Image
          style={[styles.icon, { tintColor: iconColor ? iconColor : "#fff" }]}
          source={icon}
        />
      )}

      <View>
        <Text
          style={[
            textStyles,
            {
              color:
                secondary && outlined && isPressed
                  ? ClearBackGround.Fills.Default
                  : textStyles[textStyles.length - 1].color,
            },
          ]}
        >
          {children}
        </Text>
      </View>
    </View>
  );
};

export default Infos;
