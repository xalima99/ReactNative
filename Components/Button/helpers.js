import styles from "./styles";
import { ColoredBackGround, ClearBackGround } from "../../constants/Colors";

export const getStyles = ({
  block,
  outlined,
  disabled,
  primary,
  secondary,
  link
}) => {
  const containerStyles = [styles.containerDefault];
  const textStyles = [styles.textDefault];

  if (secondary && !primary) {
    containerStyles.push(styles.containerSecondary);
    textStyles.push(styles.textSecondary);
    if (outlined) {
      containerStyles.push(styles.containerSecondaryOutlined);
      textStyles.push(styles.textSecondaryOutlined);
      if(link) containerStyles.push(styles.link);
    }
  } else {
    containerStyles.push(styles.containerPrimary);
    textStyles.push(styles.textPrimary);

    if (outlined) {
      containerStyles.push(styles.containerPrimaryOutlined);
      textStyles.push(styles.textPrimaryOutlined);
      if(link) containerStyles.push(styles.link);
    }
  }

  if (disabled && primary) {
    if (outlined) {
      containerStyles.push(styles.containerDisabledOutlined);
      if(link) containerStyles.push(styles.link);
    } else containerStyles.push(styles.containerDisabledPrimary);
    textStyles.push(styles.textDisabledPrimary);
  } else if (disabled && secondary) {
    if (outlined) {
      containerStyles.push(styles.containerDisabledOutlined);
      if(link) containerStyles.push(styles.link);
    } else containerStyles.push(styles.containerDisabledSecondary);
    textStyles.push(styles.textDisabledSecondary);
  }

  if (block) {
    containerStyles.unshift(styles.block);
  }

  // console.log(textStyles)
  return { containerStyles, textStyles };
};

export const requiredPropsCheck = (props, propName, componentName) => {
  if (!props.primary && !props.secondary) {
    return new Error(
      `One of 'primary' or 'secondary' is required by '${componentName}' component.`
    );
  }
};

export const getPressedStateColor = (
  primary,
  secondary,
  outlined,
  link,
  disabled
) => {
  if (primary) {
    if (outlined || link) return ClearBackGround.Fills.AddedBlue;
    return ClearBackGround.Fills.Actived;
  }
  //Gerer les styles onPress

  if (secondary) {
    if (outlined || link ) return ColoredBackGround.Fills.Visited;
    return ColoredBackGround.Fills.Actived;
  }

  return ClearBackGround.Fills.OceanBlueLight;
};
