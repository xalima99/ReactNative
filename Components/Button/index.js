import React, { Component } from "react";
import PropTypes from "prop-types";
import { Pressable } from "react-native";
import { getStyles, requiredPropsCheck, getPressedStateColor } from "./helpers";
import Infos from "./Infos";

class Button extends Component {
  static defaultProps = {
    outlined: false,
    disabled: false,
    block: false,
  };

  state = {
    isPressed: false,
  };

  render() {
    const {
      children,
      onPress,
      disabled,
      outlined,
      secondary,
      block,
      primary,
      icon,
      iconColor,
      link,
      style,
    } = this.props;
    const { containerStyles, textStyles } = getStyles({
      block,
      outlined,
      disabled,
      primary,
      secondary,
      link,
    });

    return (
      <Pressable
        onPress={onPress}
        disabled={disabled}
        onPressIn={() => {
          this.setState({ isPressed: true });
        }}
        onPressOut={() => {
          this.setState({ isPressed: false });
        }}
        style={({ pressed }) => {
          return [
            containerStyles,
            {
              backgroundColor: pressed
                ? getPressedStateColor(primary, secondary, outlined, disabled,link)
                : containerStyles[containerStyles.length - 1].backgroundColor,
            },
            style
          ];
        }}
      >
        <Infos
          children={children}
          icon={icon}
          iconColor={iconColor}
          textStyles={textStyles}
          secondary={secondary}
          outlined={outlined}
          isPressed={this.state.isPressed}
        />
      </Pressable>
    );
  }
}

Button.propTypes = {
  children: PropTypes.any.isRequired,
  onPress: PropTypes.func,
  outlined: PropTypes.bool,
  icon: PropTypes.any,
  iconColor: PropTypes.string,
  block: PropTypes.bool,
  primary: requiredPropsCheck,
  secondary: requiredPropsCheck,
  disabled: PropTypes.bool,
  link: PropTypes.bool,
  style: PropTypes.any
};

export default Button;
