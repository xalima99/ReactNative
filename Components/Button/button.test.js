import React from "react";
import { render } from "@testing-library/react-native";
import Button from "./index";

const button = render(<Button primary>Hello Word</Button>);

test("Button should match Snapshot", () => {
  expect(button.toJSON()).toMatchSnapshot();
});
