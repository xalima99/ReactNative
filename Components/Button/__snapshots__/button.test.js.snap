// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match Snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 6,
          "elevation": 0,
          "height": 48,
          "justifyContent": "center",
          "marginVertical": 4,
          "minHeight": 48,
          "paddingHorizontal": 16,
          "shadowColor": "#333",
          "shadowOffset": Object {
            "height": 0.3,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 0.44,
        },
        Object {
          "backgroundColor": "#00008F",
          "borderColor": "#00008F",
        },
      ],
      Object {
        "backgroundColor": "#00008F",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <Text
        style={
          Array [
            Array [
              Object {
                "color": "#fff",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ],
            Object {
              "color": "#FFFFFF",
            },
          ]
        }
      >
        Hello Word
      </Text>
    </View>
  </View>
</View>
`;
